Introduce cppjieba as a submodule for Chinese word segmentation

CrazySteve0605 opened the pull request:
Introduce [cppjieba](https://github.com/yanyiwu/cppjieba), an NLP-based Chinese tokenizer, for implementing Chinese word navigation and braille output.

[cppjieba](https://github.com/yanyiwu/cppjieba) is the C++ version of the popular Python library [jieba](https://github.com/fxsjy/jieba), offering near-identical runtime performance.

### A detialed integration analysis
#### 1. Current State of Chinese Segmentation Tools
Chinese segmentation techniques trade accuracy for runtime and memory in predictable ways:
1. Dictionary-based (fast, lower OOV coverage). Libraries such as Jieba rely on prefix trees (tries) and greedy algorithms (maximum matching / DAG). These methods are CPU-efficient and lightweight but commonly miss new or domain-specific tokens (OOV words).
2. Statistical / sequence-labeling (better context, moderate cost). Approaches using HMMs, CRFs or perceptrons (adopted by tools like THULAC, LTP, pkuseg) label characters with B/I/E/S tags to model local context and transitions. They substantially improve OOV handling and achieve high F1 on news/text benchmarks, at the cost of higher computation and memory compared with pure dictionary methods.
3. Deep learning (highest accuracy, highest cost). Transformer/BERT-based models offer superior contextual understanding and segmentation accuracy, but their inference latency and memory footprint (often hundreds of megabytes to multiple gigabytes) make them unsuitable for real-time, low-latency assistive software like NVDA.
    
### 2. NVDA-Specific Requirements
NVDA requires low latency, a small memory footprint, and reliable offline operation — which rules out heavyweight deep-learning models. At the same time, very lightweight pure dictionary- or rule-based segmenters are also unsuitable: their lower accuracy and poor OOV handling lead to frequent mis-segmentation, inaccurate speech output, and increased user confusion in real-world documents. cppjieba therefore represents a pragmatic middle ground — native C++ performance and modest resource usage combined with a hybrid strategy (dictionary lookup plus HMM-based sequence labeling) that delivers appreciably better accuracy and OOV mitigation than pure dictionary methods, while remaining far lighter and faster than Transformer/BERT approaches — making it a sensible choice for improving Chinese text handling in NVDA without imposing large runtime or memory penalties.

<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
<!-- Use Closes/Fixes/Resolves #xxx to link this PR to the issue it is responding to. -->
Related to #4075 and [a part of OSPP 2025](https://summer-ospp.ac.cn/org/prodetail/25d3e0488?list=org&navpage=org) of NVDA.

### Summary of the issue:
NVDA’s current word navigation mechanism relies on Unicode boundary rules through the [Uniscribe API](https://learn.microsoft.com/en-us/windows/win32/intl/uniscribe), which do not work well for languages such as Chinese due to the absence of explicit word delimiters.
### Description of user facing changes:
None
### Description of developer facing changes:
A tool to implement word navigation and braille output within Chinese content.
### Description of development approach:
- [x] Added cppjieba as a submodule.
- [x] Added its wrapper and building script.
### Testing strategy:
Confirm weather it can be successfully compiled and its segmentation function can be called by ctypes.
### Known issues with pull request:

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [ ] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

CrazySteve0605 committed:
Introduce cppjieba as a submodule for Chinese word segmentation

- Add `cppjieba` as a Git submodule under `third_party/cppjieba/` to provide
  robust Chinese word segmentation capabilities.
- Update `.gitmodules` to point to the official `cppjieba` repository and
  configure it to track the `master` branch.
- Update 'sconscript' to include the paths of 'cppjieba' and its dependency 'limonp'
- Modify `copying.txt` to include the `cppjieba` license (MIT) alongside the
  project’s existing license, ensuring proper attribution and compliance.
- Update documents

CrazySteve0605 committed:
Update what's new

CrazySteve0605 committed:
Add comments for building script of cppjieba and its dependency

wmhn1872265132 commented:
I downloaded the launcher built by GitHUB Actions for testing and it doesn't seem to work

cary-rowen commented:
Imo, I suggest that initial development and debugging be carried out locally, especially during the stage when functionalities are not yet operational. If extensive testing by community early users is needed, at least new features should be functional.

CrazySteve0605 commented:
> Imo, I suggest that initial development and debugging be carried out locally, especially during the stage when functionalities are not yet operational. If extensive testing by community early users is needed, at least new features should be functional.

As many parts of the code need changes, splitting the work into smaller tasks might be more efficient, as the community can review them at the same time.

seanbudd commented on nvdaHelper/local/sconscript:
```diff
@@ -113,8 +113,18 @@ localLib = env.SharedLibrary(
 		"Iphlpapi",
 		"Ws2_32",
-		"runtimeobject",
+		"runtimeobject",  
```
```suggestion
		"runtimeobject",
```

CrazySteve0605 replied:
@seanbudd Now I'm going to migrate these building scripts and directly remove changes in the whole file. Thanks for reviews.

seanbudd commented on include/readme.md:
```diff
@@ -49,3 +49,9 @@ Used in chrome system tests.
 
 Fetch latest from master.
+
+### cppjieba
```
```suggestion
### cppjieba

```

seanbudd commented on projectDocs/dev/createDevEnvironment.md:
```diff
@@ -99,6 +99,7 @@ If you aren't sure, run `git submodule update` after every git pull, merge or ch
 * [Java Access Bridge 32 bit, from Zulu Community OpenJDK build 17.0.9+8Zulu (17.46.19)](https://github.com/nvaccess/javaAccessBridge32-bin)
 * [Windows Implementation Libraries (WIL)](https://github.com/microsoft/wil/)
 * [NVDA DiffMatchPatch](https://github.com/codeofdusk/nvda_dmp)
+* [cppjieba](https://github.com/yanyiwu/cppjieba), commit '9b40903ed6cbd795367ea64f9a7d3f3bc4aa4714'
```
```suggestion
* [cppjieba - Chinese word segmentation](https://github.com/yanyiwu/cppjieba), commit '9b40903ed6cbd795367ea64f9a7d3f3bc4aa4714'
```

CrazySteve0605 committed:
Update projectDocs/dev/createDevEnvironment.md

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CrazySteve0605 committed:
Update include/readme.md

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CrazySteve0605 committed:
Remove changes in sconscript for localLIb

CrazySteve0605 committed:
add building script for cppjieba

CrazySteve0605 committed:
add JiebaSingleton wrapper and C API for NVDA segmentation

- Introduce `JiebaSingleton` class in `cppjieba.hpp`/`cppjieba.cpp` with def file under nvdaHelper/cppjieba/'
  - Inherits from `cppjieba::Jieba` and exposes a thread-safe `getOffsets()` method
  - Implements Meyers’ singleton via `getInstance()` with a private constructor
  - Deletes copy constructor, copy assignment, move constructor, and move assignment to enforce single instance
- Add C-style API in the same module:
  - `int initJieba()` to force singleton initialization
  - `int segmentOffsets(const char* text, int** charOffsets, int* outLen)` to perform segmentation and return character offsets
  - `void freeOffsets(int* ptr)` to release allocated offset buffer

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
Update GitHub action workflow to fetch cppjieba's submodule

- Change 'submodules' in 'jobs - buildNVDA - Build NVDA - Checkout NVDA' from 'true' to 'recursive' to ensure cppjieba's submodule is fetched.
- This will cause the submodule of sonic to be fetched as well, which seems currently unused.

CrazySteve0605 committed:
Update .gitignore for cppjieba

CrazySteve0605 committed:
Update building and setup script for cppjieba's dicts installation

gerald-hartig commented on nvdaHelper/cppjieba/cppjieba.cpp:
```diff
@@ -0,0 +1,85 @@
+URL: http://www.nvda-project.org/
+Copyright 2025 NV Access Limited, Wang Chong.
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License version 2.0, as published by
+	the Free Software Foundation.
```
```suggestion
it under the terms of the GNU General Public License as published by the Free Software Foundation; 
either version 2 of the License, or (at your option) any later version.
```

CrazySteve0605 replied:
Thanks for your reviews. I think I should directly replace this with the copyright header in documentation as @seanbudd suggested.

gerald-hartig commented on nvdaHelper/cppjieba/cppjieba.hpp:
```diff
@@ -0,0 +1,71 @@
+URL: http://www.nvda-project.org/
+Copyright 2025 NV Access Limited, Wang Chong.
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License version 2.0, as published by
+	the Free Software Foundation.
```
```suggestion
	it under the terms of the GNU General Public License as published by the Free Software Foundation; 
	either version 2 of the License, or (at your option) any later version.
```

seanbudd commented on include/readme.md:
```diff
@@ -49,3 +49,10 @@ Used in chrome system tests.
+
+### cppjieba
+
+[cppjieba](https://github.com/yanyiwu/cppjieba)
```
```suggestion
[cppjieba](https://github.com/yanyiwu/cppjieba).
```

seanbudd commented on nvdaHelper/cppjieba/cppjieba.cpp:
```diff
@@ -0,0 +1,85 @@
+/*
+This file is a part of the NVDA project.
```
please copy our copyright headers directly from
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/copyrightHeaders.md
and use that format

seanbudd commented on nvdaHelper/cppjieba/cppjieba.hpp:
```diff
@@ -0,0 +1,71 @@
+/*
```
same here in regards to https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/copyrightHeaders.md

seanbudd commented on nvdaHelper/cppjieba/sconscript:
```diff
@@ -0,0 +1,55 @@
+# A part of NonVisual Desktop Access (NVDA)
```
same here in regards to https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/copyrightHeaders.md

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -46,7 +46,7 @@ Please refer to [the developer guide](https://download.nvaccess.org/documentatio
 * Updated `include` dependencies:
   * detours to `9764cebcb1a75940e68fa83d6730ffaf0f669401`. (#18447, @LeonarddeR)
 * The `nvda_dmp` utility has been removed. (#18480, @codeofdusk)
-* `comInterfaces_sconscript` has been updated to make the generated files in `comInterfaces` work better with IDEs. (#17608, @gexgd0419)
```
please retain this message

CrazySteve0605 committed:
update copyright headers based on @seanbudd's suggestions

CrazySteve0605 committed:
Update include/readme.md

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CrazySteve0605 commented:
@seanbudd OK I see. I'm adding more details to make the main comment more informative. Instead of extending 'TextInfos', I'm creating a separate 'WordSegmentationStrategy' and 'WordSegmenter' that follow the strategy pattern, selecting an appropriate segmenter based on the Unicode fields present in the given text. I think this approach will be more extensible and easier to reuse in other potential components, such as braille output. I also plan to submit it as my next PR.

seanbudd commented:
That approach sounds good to me - as long as there is an easy way to test it and the initial implementation is not too complex so we get an idea of the design

cary-rowen commented:
Hi @CrazySteve0605,

I just wanted to confirm that this series of work will cover the following cases:
1. Moving by words using Ctrl+Left/Right Arrows.
2. Object Text Review.
3. Braille output.

CrazySteve0605 commented:
> Hi @CrazySteve0605,
> 
> I just wanted to confirm that this series of work will cover the following cases:
> 
>     1. Moving by words using Ctrl+Left/Right Arrows.
> 
>     2. Object Text Review.
> 
>     3. Braille output.

Hi @cary-rowen, thanks for confirmation. Things have been going that way. Word navigation via both caret (ctrl+left/right arrow by default) and review curser (numpad 4/6 by default) will remain as-is.

LeonarddeR commented:
@CrazySteve0605 Just to make sure, cppjieba only works for Chinese, right? I'm asking because a bug was discovered in uniscribe code I wrote, see #18722.

seanbudd commented:
Yes - I think we should just go with the inbuilt ICU, assuming it is fit for purpose and works well for Chinese segmentation

CrazySteve0605 commented:
> Yes - I think we should just go with the inbuilt ICU, assuming it is fit for purpose and works well for Chinese segmentation

Hi @seanbudd 
I've just updated the PR budy with some related information, and quote it as follow

> ### A detialed integration analysis

> #### 1. Current State of Chinese Segmentation Tools
> 
> Chinese segmentation techniques trade accuracy for runtime and memory in predictable ways:
> 
>     1. Dictionary-based (fast, lower OOV coverage). Libraries such as Jieba rely on prefix trees (tries) and greedy algorithms (maximum matching / DAG). These methods are CPU-efficient and lightweight but commonly miss new or domain-specific tokens (OOV words).
> 
>     2. Statistical / sequence-labeling (better context, moderate cost). Approaches using HMMs, CRFs or perceptrons (adopted by tools like THULAC, LTP, pkuseg) label characters with B/I/E/S tags to model local context and transitions. They substantially improve OOV handling and achieve high F1 on news/text benchmarks, at the cost of higher computation and memory compared with pure dictionary methods.
> 
>     3. Deep learning (highest accuracy, highest cost). Transformer/BERT-based models offer superior contextual understanding and segmentation accuracy, but their inference latency and memory footprint (often hundreds of megabytes to multiple gigabytes) make them unsuitable for real-time, low-latency assistive software like NVDA.
> 
> 
> ### 2. NVDA-Specific Requirements
> 
> NVDA requires low latency, a small memory footprint, and reliable offline operation — which rules out heavyweight deep-learning models. At the same time, very lightweight pure dictionary- or rule-based segmenters are also unsuitable: their lower accuracy and poor OOV handling lead to frequent mis-segmentation, inaccurate speech output, and increased user confusion in real-world documents. cppjieba therefore represents a pragmatic middle ground — native C++ performance and modest resource usage combined with a hybrid strategy (dictionary lookup plus HMM-based sequence labeling) that delivers appreciably better accuracy and OOV mitigation than pure dictionary methods, while remaining far lighter and faster than Transformer/BERT approaches — making it a sensible choice for improving Chinese text handling in NVDA without imposing large runtime or memory penalties.

As I learned, cppjieba implements a hybrid approach (dictionary lookup with an HMM-based backoff — effectively tiers 1 & 2), which gives stronger Chinese-specific segmentation and much better OOV handling for names, technical terms and domain vocabulary while remaining lightweight. ICU, by contrast, is primarily a language-agnostic, rule/dictionary-based segmenter (tier 1) that’s excellent for broad multilingual coverage and avoiding extra dependencies, but it typically delivers lower segmentation accuracy on Chinese text. In short: if the priority is accurate Chinese segmentation in real-world documents, cppjieba is the better-fit compromise between accuracy and resource use; if the priority is to rely solely on a built-in, language-agnostic solution with no added dependencies, ICU is an acceptable choice — just expect reduced precision on Chinese-specific content.

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
add wrappers for user dict management

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
simplify the initialization of cppjieba

drop off initialization of the keyword extractor which we don't need

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
update cppjieba to the latest commit

CrazySteve0605 committed:
update .gitattributes for .hpp header files

CrazySteve0605 committed:
simplify helper of `cppjieba`

* turn to build-in `Word` structure
* remove some items we don't use

CrazySteve0605 committed:
update changelog

CrazySteve0605 committed:
update building script

CrazySteve0605 committed:
revert installing script

CrazySteve0605 committed:
fix building script

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
update helper of `coojieba`

michaelDCurran commented:
this pr is pretty much good. Just two small things in my last review. 
We have decided that this and any other related PRs for this OSSP project will be merged into a special try-wordSegmentation-staging branch rather than master. So that all PRs can be reviewed and merged in to one working branch, which will then if successful, be merged to master all at the same time. Therefore I have switched the base branch of this pr to try-wordSegmentation-staging accordingly.

seanbudd commented on projectDocs/dev/createDevEnvironment.md:
```diff
@@ -100,7 +100,7 @@ If you aren't sure, run `git submodule update` after every git pull, merge or ch
 * [Java Access Bridge 32 bit, from Zulu Community OpenJDK build 17.0.9+8Zulu (17.46.19)](https://github.com/nvaccess/javaAccessBridge32-bin)
   * We are in the process of switching to: Java Access Bridge 64 bit, from Zulu Community OpenJDK build 17.0.16+8Zulu (17.60.17)
 * [Windows Implementation Libraries (WIL)](https://github.com/microsoft/wil/)
-* [NVDA DiffMatchPatch](https://github.com/codeofdusk/nvda_dmp)
```
We did remove this, it just should go in another PR

CrazySteve0605 commented on nvdaHelper/cppjieba/cppjieba.hpp:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  static string getPath(const string& path, const string& default_file) {
+    if (path.empty()) {
```
That's true. This also caused NVDA failed to launch on other tester's PC but I didn't notice that it is at all. Thanks.

CrazySteve0605 committed:
Merge branch 'master' into integrateCPPJieba

CrazySteve0605 committed:
Revert "Update projectDocs/dev/createDevEnvironment.md"

This reverts commit 06070c12f7261ddac8dbd223e6d561950cad500d.

CrazySteve0605 committed:
avoid using compilation time path

CrazySteve0605 committed:
Update .gitattributes

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

michaelDCurran committed:
Merge branch 'try-chineseWordSegmentation-staging' into integrateCPPJieba

michaelDCurran committed:
Merge branch 'try-chineseWordSegmentation-staging' into integrateCPPJieba

michaelDCurran commented:
Please note, that I incorrectly had merged this as a squash commit, which caused problems for dependent PRs such as #18735.
 I have now reverted that squash commit and  merged this PR's branch unsquashed as commit 093a825fb3f60d48c6b8ebe2
